modules/reindex/src/main/java/org/elasticsearch/index/reindex/remote/RemoteRequestBuilders.java

```diff
@@ -0,0 +1,171 @@
+                    for (SortBuilder<?> sort : searchRequest.source().sorts()) {
+                        if (sort instanceof FieldSortBuilder) {
+                            FieldSortBuilder f = (FieldSortBuilder) sort;
+                            if (f.getFieldName().equals(FieldSortBuilder.DOC_FIELD_NAME)) {
+                                useScan = true;
+                                break;
+                            }
+                        }
+                    }
+                }
+                if (useScan) {
+                    params.put("search_type", "scan");
+                } else {
+                    StringBuilder sorts = new StringBuilder(sortToUri(searchRequest.source().sorts().get(0)));
+                    for (int i = 1; i < searchRequest.source().sorts().size(); i++) {
+                        sorts.append(',').append(sortToUri(searchRequest.source().sorts().get(i)));
+                    }
+                    params.put("sorts", sorts.toString());
+                }
+            }
+            if (searchRequest.source().storedFields() != null && false == searchRequest.source().storedFields().isEmpty()) {
+                StringBuilder fields = new StringBuilder(searchRequest.source().storedFields().get(0));
+                for (int i = 1; i < searchRequest.source().storedFields().size(); i++) {
+                   fields.append(',').append(searchRequest.source().storedFields().get(i));
+                }
+                String storedFieldsParamName = remoteVersion.before(Version.V_5_0_0_alpha4) ? "fields" : "stored_fields";
+                params.put(storedFieldsParamName, fields.toString());
+            }
+            return params;
+        }
+
+        static HttpEntity entity(BytesReference query) {
+            try (XContentBuilder entity = JsonXContent.contentBuilder();
+                    XContentParser queryParser = XContentHelper.createParser(query)) {
+                entity.startObject();
+                entity.field("query");
+                /*
+                 * We're intentionally a bit paranoid here - copying the query as xcontent rather than writing a raw field. We don't want
+                 * poorly written queries to escape. Ever.
+                 */
+                entity.copyCurrentStructure(queryParser);
+                XContentParser.Token shouldBeEof = queryParser.nextToken();
+                if (shouldBeEof != null) {
+                    throw new RuntimeException(
+                            "query was more than a single object. This first token after the object is [" + shouldBeEof + "]");
+                }
+                entity.endObject();
+                BytesReference bytes = entity.bytes();
+                return new ByteArrayEntity(bytes.array(), bytes.arrayOffset(), bytes.length(), ContentType.APPLICATION_JSON);
+            } catch (IOException e) {
+                throw new RuntimeException("Unexpected error building entity", e);
```

Comment:
also we usually lowercase all the log messages and exception messages too (minor nit)